README.md
# pydoautomator

[![Downloads](https://pepy.tech/badge/pydoautomator/month)](https://pepy.tech/project/pydoautomator)
[![codecov](https://codecov.io/gh/christian-hawk/pydoautomator/branch/master/graph/badge.svg)](https://codecov.io/gh/christian-hawk/pydoautomator)
[![DeepSource](https://deepsource.io/gh/christian-hawk/pydoautomator.svg/?label=active+issues&show_trend=true)](https://deepsource.io/gh/christian-hawk/pydoautomator/?ref=repository-badge) [![Actions Status](https://github.com/christian-hawk/pydoautomator/workflows/Python/badge.svg)](https://github.com/christian-hawk/pydoautomator/actions)
[![Python Versions](https://img.shields.io/pypi/pyversions/pydoautomator.svg)](https://pypi.org/project/pydoautomator)

The Digital Ocean python automation lib

## Install

`pip install pydoautomator`

## Simple as that

``` python
from pydoautomator import Automator, Droplet
digital_ocean_token = 'my-digital-ocean-api-token'

aut = Automator(digital_ocean_token)

droplet_data = {
    "name" : "t1.techno24x7.com",
    "region" : "nyc1",
    "size" : "s-8vcpu-16gb",
    "image" : 68259296, # snapshot id
    "ssh_keys" : [27410347, 27608055, 27590881],
    "private_networking" : True,
    "vpc_uuid" : "47e5c00a-2b23-4dac-bed4-0e44659941f3",
    "monitoring" : True
    "tags" : ["tests"]
}

droplet = Droplet(**droplet_data)

aut.create_droplet_from_snapshot(droplet)

```

### Assign floating ip to droplet

```python
droplet_id = 152412424
floating_ip = '164.90.252.72'
action_status = aut.assign_floating_ip_to_droplet(floating_ip, droplet_id)

if action_status == 'completed':
    print('floating_ip assigned to droplet!')
```

### Create droplet and assign floating ip as soon as droplet created

```python
digital_ocean_token = 'my-super-cool-digital-ocean-api-token'

aut = Automator(digital_ocean_token)

droplet_data = {
    "name": "t1.techno24x7.com",
    "region": "nyc1",
    "size": "s-8vcpu-16gb",
    "image": 70649304, # snapshot id
    "ssh_keys": [27410347, 27608055, 27590881],
    "private_networking": True,
    "vpc_uuid": "47e5c00a-2b23-4dac-bed4-0e44659941f3",
    "monitoring": True
}

droplet = Droplet(**droplet_data)

droplet_id = aut.create_droplet_from_snapshot(droplet)


floating_ip = '164.90.252.72'
action_status = aut.assign_floating_ip_to_droplet(floating_ip, droplet_id)

if action_status == 'completed':
    print('floating_ip assigned to droplet!')
```

### Shutdown / turnoff droplet

```python
droplet_id = 123456
aut.turnoff_droplet(droplet_id)
```

### Destroy droplet by id

```python
droplet_id = 123456
aut.destroy_droplet(droplet_id)
```

### Get all droplets

```python
droplets_list = aut.get_all_droplets(droplet_id)
```


CHANGELOG.md
## 0.8.2 (2020-10-25)

### Refactor

- removed unused imports

## 0.8.1 (2020-10-24)

### Refactor

- **.gitignore**: add .vscode

### Fix

- **droplet.py**: vpc_uuid is now optional

## 0.8.0 (2020-10-24)

- Add auto-bump

## 0.7.0 (2020-10-21)

### Feat

- **droplet.py**: add tags field to droplet model

## 0.6.0 (2020-10-20)

### Feat

- **automator.py**: Destroy droplet feature

### Fix

- **pyproject.toml**: Fix importlib module required for some python versions (not native)

## 0.5.0 (2020-10-19)

### Feat

- **automator.py**: turnoff_droplet

### Refactor

- **test_assign_floating_ip.py**: fixed typo

### Fix

- **test_automator.py**: Stash mocks after each use

## 0.4.0 (2020-10-04)

### Feat

- **Automator**: Add get_all_droplets to Automator

## 0.3.1 (2020-09-29)

### Refactor

- **anti-patterns**: Fixed anti-patterns

## 0.3.0 (2020-09-29)

### Feat

- **automator.py**: assign floating ip to droplet

## 0.2.0 (2020-09-29)

### Feat

- **project**: Added commitizen for following commit convention

### Fix

- added .env.sh

## 0.1.3 (2020-09-28)

### Fix

- importing submodules

## 0.1.2 (2020-09-28)

### Fix

- updated version and readme
- ignoring setup.py for code coverage
- codecov yaml typo
- codecov and setup.py
- remove unused import
- Chained comparison PYL-R1716
- unused var
- fixing attrib op


tests/unit_tdd/helper.py
import json


def get_json_from_file(filename):
    with open('tests/test_data/%s' % filename) as json_file:
        return json.load(json_file)


# Test data
valid_droplet = {
    "name": "t1.techno24x7.com",
    "region": "nyc1",
    "size": "s-8vcpu-16gb",
    "image": 68259296,
    "ssh_keys": [27410347, 27608055, 27590881],
    "private_networking": True,
    "vpc_uuid": "47e5c00a-2b23-4dac-bed4-0e44659941f3",
    "monitoring": True
}

valid_droplet_created_json_response = '{"droplet":{"id":209450285,"name":"t1.techno24x7.com","memory":16384,"vcpus":8,"disk":320,"locked":false,"status":"new","kernel":null,"created_at":"2020-09-26T03:27:40Z","features":[],"backup_ids":[],"next_backup_window":null,"snapshot_ids":[],"image":{"id":68259296,"name":"t1.techno24x7.com-final-test420","distribution":"Ubuntu","slug":null,"public":false,"regions":["nyc1"],"created_at":"2020-08-11T20:38:19Z","min_disk_size":160,"type":"snapshot","size_gigabytes":5.13,"tags":[],"status":"available"},"volume_ids":[],"size":{"slug":"s-8vcpu-16gb","memory":16384,"vcpus":8,"disk":320,"transfer":6.0,"price_monthly":80.0,"price_hourly":0.11905,"regions":["ams2","ams3","blr1","fra1","lon1","nyc1","nyc2","nyc3","sfo1","sfo2","sfo3","sgp1","tor1"],"available":true},"size_slug":"s-8vcpu-16gb","networks":{"v4":[],"v6":[]},"region":{"name":"New York 1","slug":"nyc1","features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"available":true,"sizes":["s-1vcpu-1gb","512mb","s-1vcpu-2gb","1gb","s-3vcpu-1gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","2gb","s-4vcpu-8gb","m-1vcpu-8gb","c-2","4gb","c2-2vcpu-4gb","g-2vcpu-8gb","gd-2vcpu-8gb","m-16gb","s-8vcpu-16gb","s-6vcpu-16gb","c-4","8gb","c2-4vpcu-8gb","m-2vcpu-16gb","m3-2vcpu-16gb","g-4vcpu-16gb","gd-4vcpu-16gb","m6-2vcpu-16gb","m-32gb","s-8vcpu-32gb","c-8","c2-8vpcu-16gb","m-4vcpu-32gb","m3-4vcpu-32gb","g-8vcpu-32gb","s-12vcpu-48gb","gd-8vcpu-32gb","m6-4vcpu-32gb","m-64gb","s-16vcpu-64gb","c-16","32gb","c2-16vcpu-32gb","m-8vcpu-64gb","m3-8vcpu-64gb","g-16vcpu-64gb","s-20vcpu-96gb","48gb","gd-16vcpu-64gb","m6-8vcpu-64gb","m-128gb","s-24vcpu-128gb","64gb","c2-32vpcu-64gb","m-16vcpu-128gb","m3-16vcpu-128gb","s-32vcpu-192gb","m-24vcpu-192gb","m6-16vcpu-128gb","m3-24vcpu-192gb","m6-24vcpu-192gb"]},"tags":[]},"links":{"actions":[{"id":1030573508,"rel":"create","href":"https://api.digitalocean.com/v2/actions/1030573508"}]}}'

valid_droplet_created_response = json.loads(
    valid_droplet_created_json_response)

valid_creation_in_progress_json_resp = '{"action":{"id":1030573508,"status":"in-progress","type":"create","started_at":"2020-09-26T03:27:40Z","completed_at":null,"resource_id":209450285,"resource_type":"droplet","region":{"name":"New York 1","slug":"nyc1","features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"available":true,"sizes":["c-2","c-4","m-1vcpu-8gb","m-16gb","m-32gb","m-64gb","m-128gb","512mb","1gb","2gb","4gb","8gb","32gb","48gb","64gb","s-1vcpu-1gb","s-3vcpu-1gb","s-1vcpu-2gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","s-4vcpu-8gb","s-6vcpu-16gb","s-8vcpu-32gb","g-2vcpu-8gb","m-2vcpu-16gb","m3-2vcpu-16gb","m6-2vcpu-16gb","s-12vcpu-48gb","s-16vcpu-64gb","s-20vcpu-96gb","s-24vcpu-128gb","s-32vcpu-192gb","gd-2vcpu-8gb","s-8vcpu-16gb","c2-2vcpu-4gb","c2-4vpcu-8gb","c2-8vpcu-16gb","c2-16vcpu-32gb","c2-32vpcu-64gb"]},"region_slug":"nyc1"}}'

valid_creation_in_progress_response = json.loads(
    valid_creation_in_progress_json_resp)

valid_creation_completed_json_resp = '{"action":{"id":1030573508,"status":"completed","type":"create","started_at":"2020-09-26T03:27:40Z","completed_at":"2020-09-26T03:33:44Z","resource_id":209450285,"resource_type":"droplet","region":{"name":"New York 1","slug":"nyc1","features":["private_networking","backups","ipv6","metadata","install_agent","storage","image_transfer"],"available":true,"sizes":["c-2","c-4","m-1vcpu-8gb","m-16gb","m-32gb","m-64gb","m-128gb","512mb","1gb","2gb","4gb","8gb","32gb","48gb","64gb","s-1vcpu-1gb","s-3vcpu-1gb","s-1vcpu-2gb","s-2vcpu-2gb","s-1vcpu-3gb","s-2vcpu-4gb","s-4vcpu-8gb","s-6vcpu-16gb","s-8vcpu-32gb","g-2vcpu-8gb","m-2vcpu-16gb","m3-2vcpu-16gb","m6-2vcpu-16gb","s-12vcpu-48gb","s-16vcpu-64gb","s-20vcpu-96gb","s-24vcpu-128gb","s-32vcpu-192gb","gd-2vcpu-8gb","s-8vcpu-16gb","c2-2vcpu-4gb","c2-4vpcu-8gb","c2-8vpcu-16gb","c2-16vcpu-32gb","c2-32vpcu-64gb"]},"region_slug":"nyc1"}}'

valid_creation_completed_response = json.loads(
    valid_creation_completed_json_resp)


tests/unit_tdd/api_mocks.py
import responses
from .helper import get_json_from_file


# mock api responses


def error_404_response(url):
    """
    Respond a 404 with dummy json
    works with POST and DELETE
    """
    responses.add(
        responses.POST,
        url=url,
        json={'any_key': 'any_value'},
        status=404
    )
    responses.add(
        responses.DELETE,
        url=url,
        json={'any_key': 'any_value'},
        status=404
    )


def destroy_droplet_success_response(droplet_id: int):
    """needs to be used with @responses.activate"""
    responses.add(
        responses.DELETE,
        url='https://api.digitalocean.com/v2/droplets/%s' % droplet_id,
        json="",
        status=204
    )


def shutdown_droplet_success_response(droplet_id: int):
    """needs to be used with @responses.activate"""
    response_action_data = {
        "action": {
            "id": 1047023793,
            "status": "in-progress",
            "type": "shutdown",
            "started_at": "2020-10-18T19:20:00Z",
            "completed_at": None,
            "resource_id": droplet_id,
            "resource_type": "droplet",
            "region": {
                "name": "New York 1",
                "slug": "nyc1",
                "features": [
                    "private_networking",
                    "backups",
                    "ipv6",
                    "metadata",
                    "install_agent",
                    "storage",
                    "image_transfer"
                ],
                "available": True,
                "sizes": [
                    "s-1vcpu-1gb",
                    "512mb",
                    "s-1vcpu-2gb",
                    "1gb",
                    "s-3vcpu-1gb",
                    "s-2vcpu-2gb",
                    "s-1vcpu-3gb",
                    "s-2vcpu-4gb",
                    "2gb",
                    "s-4vcpu-8gb",
                    "m-1vcpu-8gb",
                    "c-2",
                    "4gb",
                    "c2-2vcpu-4gb",
                    "g-2vcpu-8gb",
                    "gd-2vcpu-8gb",
                    "m-16gb",
                    "s-8vcpu-16gb",
                    "s-6vcpu-16gb",
                    "c-4",
                    "8gb",
                    "c2-4vpcu-8gb",
                    "m-2vcpu-16gb",
                    "m3-2vcpu-16gb",
                    "g-4vcpu-16gb",
                    "gd-4vcpu-16gb",
                    "m6-2vcpu-16gb",
                    "m-32gb",
                    "s-8vcpu-32gb",
                    "c-8",
                    "c2-8vpcu-16gb",
                    "m-4vcpu-32gb",
                    "m3-4vcpu-32gb",
                    "g-8vcpu-32gb",
                    "s-12vcpu-48gb",
                    "gd-8vcpu-32gb",
                    "m6-4vcpu-32gb",
                    "m-64gb",
                    "s-16vcpu-64gb",
                    "c-16",
                    "32gb",
                    "c2-16vcpu-32gb",
                    "m-8vcpu-64gb",
                    "m3-8vcpu-64gb",
                    "g-16vcpu-64gb",
                    "s-20vcpu-96gb",
                    "48gb",
                    "gd-16vcpu-64gb",
                    "m6-8vcpu-64gb",
                    "m-128gb",
                    "s-24vcpu-128gb",
                    "c-32",
                    "64gb",
                    "c2-32vpcu-64gb",
                    "m-16vcpu-128gb",
                    "m3-16vcpu-128gb",
                    "g-32vcpu-128gb",
                    "s-32vcpu-192gb",
                    "m-24vcpu-192gb",
                    "m6-16vcpu-128gb",
                    "m3-24vcpu-192gb",
                    "m6-24vcpu-192gb",
                    "m3-32vcpu-256gb",
                    "m6-32vcpu-256gb"
                ]
            },
            "region_slug": "nyc1"
        }
    }
    responses.add(
        responses.POST,
        url='https://api.digitalocean.com/v2/droplets/%s/actions' % droplet_id,
        json=response_action_data,
        status=201
    )


def assign_floating_ip_success_response(floating_ip: str, droplet_id: int):
    """needs to be used with @responses.activate
    Successfull responses for params

    :param floating_ip: floating ip to be sucessfully assigned
    :type floating_ip: str
    :param droplet_id: droplet id to sucessfully receive fip
    :type droplet_id: int
    """

    response_action_data = {
        "action": {
            "id": 1032896014,
            "status": "in-progress",
            "type": "assign_ip",
            "started_at": "2020-09-29T18:21:38Z",
            "completed_at": None,
            "resource_id": 2757426197,
            "resource_type": "floating_ip",
            "region": {
                "name": "New York 1",
                "slug": "nyc1",
                "features": [
                    "private_networking",
                    "backups",
                    "ipv6",
                    "metadata",
                    "install_agent",
                    "storage",
                    "image_transfer"
                ],
                "available": True,
                "sizes": [
                    "c-2",
                    "c-4",
                    "m-1vcpu-8gb",
                    "m-16gb",
                    "m-32gb",
                    "m-64gb",
                    "m-128gb",
                    "512mb",
                    "1gb",
                    "2gb",
                    "4gb",
                    "8gb",
                    "32gb",
                    "48gb",
                    "64gb",
                    "s-1vcpu-1gb",
                    "s-3vcpu-1gb",
                    "s-1vcpu-2gb",
                    "s-2vcpu-2gb",
                    "s-1vcpu-3gb",
                    "s-2vcpu-4gb",
                    "s-4vcpu-8gb",
                    "s-6vcpu-16gb",
                    "s-8vcpu-32gb",
                    "g-2vcpu-8gb",
                    "m-2vcpu-16gb",
                    "m3-2vcpu-16gb",
                    "m6-2vcpu-16gb",
                    "s-12vcpu-48gb",
                    "s-16vcpu-64gb",
                    "s-20vcpu-96gb",
                    "s-24vcpu-128gb",
                    "s-32vcpu-192gb",
                    "gd-2vcpu-8gb",
                    "s-8vcpu-16gb",
                    "c2-2vcpu-4gb",
                    "c2-4vpcu-8gb",
                    "c2-8vpcu-16gb",
                    "c2-16vcpu-32gb",
                    "c2-32vpcu-64gb"
                ]
            },
            "region_slug": "nyc1"
        }
    }

    responses.add(
        responses.POST,
        url='https://api.digitalocean.com/v2/floating_ips/%s/actions' % floating_ip,
        json=response_action_data,
        status=201
    )


def existing_droplet_response():
    """needs to be used with @responses.activate
    """
    responses.add(
        responses.GET,
        url='https://api.digitalocean.com/v2/droplets/209808688',
        json={
            "droplet": {
                "id": 209808688,
                "name": "t1.techno24x7.com",
                "memory": 16384,
                "vcpus": 8,
                "disk": 320,
                "locked": False,
                "status": "active",
                "kernel": None,
                "created_at": "2020-09-29T00:59:22Z",
                "features": [
                    "monitoring",
                    "private_networking"
                ],
                "backup_ids": [],
                "next_backup_window": None,
                "snapshot_ids": [],
                "image": {
                    "id": 70649304,
                    "name": "t1.techno24x7.com-latest-dev",
                    "distribution": "Ubuntu",
                    "slug": None,
                    "public": False,
                    "regions": [
                        "nyc1"
                    ],
                    "created_at": "2020-09-24T22:54:22Z",
                    "min_disk_size": 320,
                    "type": "snapshot",
                    "size_gigabytes": 1.92,
                    "tags": [],
                    "status": "available"
                },
                "volume_ids": [],
                "size": {
                    "slug": "s-8vcpu-16gb",
                    "memory": 16384,
                    "vcpus": 8,
                    "disk": 320,
                    "transfer": 6.0,
                    "price_monthly": 80.0,
                    "price_hourly": 0.11905,
                    "regions": [
                        "ams2",
                        "ams3",
                        "blr1",
                        "fra1",
                        "lon1",
                        "nyc1",
                        "nyc2",
                        "nyc3",
                        "sfo1",
                        "sfo2",
                        "sfo3",
                        "sgp1",
                        "tor1"
                    ],
                    "available": True
                },
                "size_slug": "s-8vcpu-16gb",
                "networks": {
                    "v4": [
                        {
                            "ip_address": "10.116.0.2",
                            "netmask": "255.255.240.0",
                            "gateway": "<nil>",
                            "type": "private"
                        },
                        {
                            "ip_address": "161.35.110.172",
                            "netmask": "255.255.240.0",
                            "gateway": "161.35.96.1",
                            "type": "public"
                        }
                    ],
                    "v6": []
                },
                "region": {
                    "name": "New York 1",
                    "slug": "nyc1",
                    "features": [
                        "private_networking",
                        "backups",
                        "ipv6",
                        "metadata",
                        "install_agent",
                        "storage",
                        "image_transfer"
                    ],
                    "available": True,
                    "sizes": [
                        "s-1vcpu-1gb",
                        "512mb",
                        "s-1vcpu-2gb",
                        "1gb",
                        "s-3vcpu-1gb",
                        "s-2vcpu-2gb",
                        "s-1vcpu-3gb",
                        "s-2vcpu-4gb",
                        "2gb",
                        "s-4vcpu-8gb",
                        "m-1vcpu-8gb",
                        "c-2",
                        "4gb",
                        "c2-2vcpu-4gb",
                        "g-2vcpu-8gb",
                        "gd-2vcpu-8gb",
                        "m-16gb",
                        "s-8vcpu-16gb",
                        "s-6vcpu-16gb",
                        "c-4",
                        "8gb",
                        "c2-4vpcu-8gb",
                        "m-2vcpu-16gb",
                        "m3-2vcpu-16gb",
                        "g-4vcpu-16gb",
                        "gd-4vcpu-16gb",
                        "m6-2vcpu-16gb",
                        "m-32gb",
                        "s-8vcpu-32gb",
                        "c-8",
                        "c2-8vpcu-16gb",
                        "m-4vcpu-32gb",
                        "m3-4vcpu-32gb",
                        "g-8vcpu-32gb",
                        "s-12vcpu-48gb",
                        "gd-8vcpu-32gb",
                        "m6-4vcpu-32gb",
                        "m-64gb",
                        "s-16vcpu-64gb",
                        "c-16",
                        "32gb",
                        "c2-16vcpu-32gb",
                        "m-8vcpu-64gb",
                        "m3-8vcpu-64gb",
                        "g-16vcpu-64gb",
                        "s-20vcpu-96gb",
                        "48gb",
                        "gd-16vcpu-64gb",
                        "m6-8vcpu-64gb",
                        "m-128gb",
                        "s-24vcpu-128gb",
                        "c-32",
                        "64gb",
                        "c2-32vpcu-64gb",
                        "m-16vcpu-128gb",
                        "m3-16vcpu-128gb",
                        "s-32vcpu-192gb",
                        "m-24vcpu-192gb",
                        "m6-16vcpu-128gb",
                        "m3-24vcpu-192gb",
                        "m6-24vcpu-192gb"
                    ]
                },
                "tags": [],
                "vpc_uuid": "47e5c00a-2b23-4dac-bed4-0e44659941f3"
            }
        },
        status=200)


def list_droplets_response():
    responses.reset()
    responses.add(
        responses.GET,
        url='https://api.digitalocean.com/v2/droplets',
        json=get_json_from_file('list_droplets_page_1.json'),
        status=200
    )
    responses.add(
        responses.GET,
        url='https://api.digitalocean.com/v2/droplets?page=2&per_page=20',
        json=get_json_from_file('list_droplets_page_2.json'),
        status=200
    )
    responses.add(
        responses.GET,
        url='https://api.digitalocean.com/v2/droplets?page=3&per_page=20',
        json=get_json_from_file('list_droplets_page_3.json'),
        status=200
    )


tests/unit_tdd/test_errors.py
import pytest
from pydoautomator import errors
import inspect
from pydoautomator.errors import DropletCreationError, FloatingIpAssignmentError, TurnoffDropletError, DestroyDropletError


def test_if_submodule_automator_exists():
    try:
        import pydoautomator.errors as errors  # skipcq: PYL-W0611
    except ModuleNotFoundError:
        pytest.fail('submodule errors does NOT exist')


def test_droplet_creation_error_should_exist():
    assert hasattr(errors, 'DropletCreationError')


def test_droplet_creation_error_should_be_a_class():
    assert inspect.isclass(errors.DropletCreationError)


def test_droplet_creation_error_should_be_exception_subclass():
    assert issubclass(errors.DropletCreationError, Exception)


def test_droplet_creation_error_should_have_msg():
    expected_msg = 'Digital Ocean API returned errored'
    with pytest.raises(DropletCreationError, match=expected_msg):
        raise DropletCreationError


def test_floating_ip_assignment_error_should_exist():
    assert hasattr(errors, 'FloatingIpAssignmentError')


def test_floating_ip_assignment_error_should_be_a_class():
    assert inspect.isclass(errors.FloatingIpAssignmentError)


def test_floating_ip_assignment_error_should_be_exception_subclass():
    assert issubclass(errors.FloatingIpAssignmentError, Exception)


def test_floating_ip_assignment_error_should_have_msg():
    msg = {'error': 'yeah right'}
    expected_msg = 'Error on assigning floating ip. Response data is: ' + \
        str(msg)

    with pytest.raises(FloatingIpAssignmentError, match=expected_msg):
        raise FloatingIpAssignmentError(msg)


def test_turnoff_droplet_error_should_exist():
    assert hasattr(errors, 'TurnoffDropletError')


def test_turnoff_droplet_error_should_be_a_class():
    assert inspect.isclass(errors.TurnoffDropletError)


def test_turnoff_droplet_error_should_be_exception_subclass():
    assert issubclass(errors.TurnoffDropletError, Exception)


def test_turnoff_droplet_should_have_msg():
    msg = {'error': 'this error message was sent'}
    expected_msg = 'Error shutting down droplet. Response data is: ' + \
        str(msg)

    with pytest.raises(TurnoffDropletError, match=expected_msg):
        raise TurnoffDropletError(msg)


def test_destroy_droplet_error_should_exist():
    assert hasattr(errors, 'DestroyDropletError')


def test_destroy_droplet_error_should_be_a_class():
    assert inspect.isclass(errors.DestroyDropletError)


def test_destroy_droplet_error_should_be_exception_subclass():
    assert issubclass(errors.DestroyDropletError, Exception)


def test_destroy_droplet_should_have_msg():
    msg = {'error': 'this error message was sent'}
    expected_msg = 'Error destroying droplet. Response data is: ' + \
        str(msg)

    with pytest.raises(DestroyDropletError, match=expected_msg):
        raise DestroyDropletError(msg)


pydoautomator/automator.py
from pydoautomator.droplet import Droplet
from pydoautomator.adapters import ApiAdapter
import asyncio
from pydoautomator.errors import DropletCreationError, FloatingIpAssignmentError, TurnoffDropletError, DestroyDropletError


class Automator:
    """Here are the main methods / functions
    Call this class passing the Digital Ocean TOKEN.

    Raises:
        TurnoffDropletError: Error while shutting down droplet
        FloatingIpAssignmentError: Error while assigning floating ip
        DestroyDropletError: Error while destroying droplet
        DropletCreationError: Error while creating droplet

    """

    do_token = ''

    requests = ''
    __base_url = 'https://api.digitalocean.com/v2'

    def __init__(self, do_token: str):
        """constructor

        Args:
            do_token (str): digital ocean token
        """

        self.do_token = do_token
        self.api_adapter = ApiAdapter(self.do_token)
        self.requests = self.api_adapter.requests
        self.__base_url = 'https://api.digitalocean.com/v2'

    def turnoff_droplet(self, droplet_id: int) -> str:
        """Turnoff / Shutdown Droplet

        Args:
            droplet_id (int): droplet id

        Raises:
            TurnoffDropletError: Error if droplet is not shutdown

        Returns:
            str: `completed`
        """

        headers = {'Content-Type': 'application/json'}
        data = {'type': 'shutdown'}

        response = self.requests.post(
            self.__base_url+'/droplets/'+str(droplet_id)+'/actions',
            headers=headers,
            json=data
        )

        if response.status_code != 201:
            raise TurnoffDropletError(response.json())

        action_id = response.json()['action']['id']

        loop2 = asyncio.get_event_loop()
        loop2.run_until_complete(
            self.__wait_till_action_complete(action_id)
        )
        return "completed"

    def get_all_droplets(self) -> list:
        """Get all droplets from Digital Ocean

        Returns:
            list: list of droplets (dict)
        """

        response = self.requests.get(self.__base_url+'/droplets')
        droplets = response.json()['droplets']

        while 'next' in response.json()['links']['pages']:
            response = self.requests.get(
                response.json()['links']['pages']['next'])
            droplets = droplets + response.json()['droplets']

        return droplets

    def assign_floating_ip_to_droplet(self, floating_ip: str, droplet_id: int) -> str:
        """Assigns a floating ip to a droplet

        Args:
            floating_ip (str): Floating IP (i.e.: `125.68.75.2`)
            droplet_id (int): The droplet ID

        Raises:
            FloatingIpAssignmentError: if not completed

        Returns:
            str: return `completed` if completed
        """

        url = self.__base_url + '/floating_ips/' + floating_ip + '/actions'
        data = {
            "type": "assign",
            "droplet_id": droplet_id
        }

        headers = {'Content-Type': 'application/json'}
        response = self.requests.post(url, json=data, headers=headers)

        if response.status_code != 201:
            raise FloatingIpAssignmentError(response.json())

        action_id = response.json()['action']['id']

        loop2 = asyncio.get_event_loop()
        loop2.run_until_complete(
            self.__wait_till_action_complete(action_id)
        )
        if response.status_code == 201:
            return 'completed'

    def create_droplet_from_snapshot(self, droplet: Droplet) -> int:
        """Creates droplet from snapshot/image

        Args:
            droplet (Droplet): Droplet object

        Returns:
            int: `droplet_id` from created droplet
        """

        headers = {'Content-Type': 'application/json'}
        created_droplet = self.requests.post(
            self.__base_url + '/droplets', data=droplet.json(), headers=headers)

        action_id = created_droplet.json()['links']['actions'][0]['id']
        loop = asyncio.get_event_loop()

        loop.run_until_complete(
            self.__wait_till_action_complete(action_id)
        )

        return created_droplet.json()['droplet']['id']

    def destroy_droplet(self, droplet_id: int) -> str:
        """Destroy / Delete droplet

        Args:
            droplet_id (int): droplet id to be destroyed

        Raises:
            DestroyDropletError: if droplet is not destroyed

        Returns:
            str: `completed` if destroyed
        """

        response = self.requests.delete(
            self.__base_url + '/droplets/' + str(droplet_id)
        )

        if response.status_code != 204:
            raise DestroyDropletError(response.json())

        return 'completed'

    async def __wait_till_action_complete(self, action_id: int) -> str:
        action_status = self.__check_action_status(action_id)

        while action_status == 'in-progress':
            await asyncio.sleep(5)
            action_status = self.__check_action_status(action_id)

        if action_status == 'errored':
            raise DropletCreationError

        return action_status

    def __check_action_status(self, action_id: int) -> str:
        """Checks digital ocean action status

        Args:
            action_id (int): Digital Ocean action id

        Returns:
            str: action status (`in-progress`, `completed`, `errored`)
        """
        print('Entered __check_action_status')

        response = self.requests.get(
            self.__base_url + '/actions/' + str(action_id)
        )
        print(response.json())

        return response.json()['action']['status']
